sampler.go

```diff
@@ -220,18 +223,29 @@ func NewGuaranteedThroughputProbabilisticSampler(
 		return nil, err
 	}
 	return &GuaranteedThroughputProbabilisticSampler{
-		probabilisticSampler: probabilisticSampler,
-		lowerBoundSampler:    NewRateLimitingSampler(lowerBound),
-		tags:                 tags,
-		samplingRate:         samplingRate,
-		lowerBound:           lowerBound,
+		probabilisticSampler:  probabilisticSampler,
+		lowerBoundSampler:     NewRateLimitingSampler(lowerBound),
+		upperBoundRateLimiter: utils.NewRateLimiter(upperBound, upperBound),
+		tags:         tags,
+		samplingRate: samplingRate,
+		lowerBound:   lowerBound,
+		upperBound:   upperBound,
 	}, nil
 }
 
 // IsSampled implements IsSampled() of Sampler.
 func (s *GuaranteedThroughputProbabilisticSampler) IsSampled(id TraceID, operation string) (bool, []Tag) {
 	if sampled, tags := s.probabilisticSampler.IsSampled(id, operation); sampled {
 		s.lowerBoundSampler.IsSampled(id, operation)
+		if !s.upperBoundRateLimiter.CheckCredit(1) {
+			// Client is oversampling, reduce the sampling probability
```

Comment:
move this to a function `reduceSamplingProbability`, and document it similar to this PR's description